cpp/src/arrow/testing/random.cc

```diff
@@ -916,14 +916,14 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(const Field& field, int64_t
 
     case Type::type::STRUCT: {
       ArrayVector child_arrays(field.type()->num_fields());
-      std::vector<std::string> field_names;
+      FieldVector child_fields;
       for (int i = 0; i < field.type()->num_fields(); i++) {
         const auto& child_field = field.type()->field(i);
         child_arrays[i] = ArrayOf(*child_field, length, alignment, memory_pool);
-        field_names.push_back(child_field->name());
+        child_fields.push_back(child_field);
       }
       return *StructArray::Make(
-          child_arrays, field_names,
+          child_arrays, child_fields,
```

Comment:
Shouldn't the original overload get the data type from `child_arrays`? I saw a comment of the current overload saying `This method does not check that field types and child array types are consistent.` It seems that we need to fix the original `Make` function as well.

Reply:
Yes, In fact I use `Generator` to generate testing data, but it `ValidationFull` fails, lol
Maybe @lidavidm can take a look?

Reply:
Sorry, what are you saying fails? I agree the proposed code is more correct.

Reply:
Ok, but I don't really get what this thread is trying to discuss.

@wgtmac note that types do not carry nullability info. So we need to use fields here, this isn't a problem with Make.

Reply:
I just realized that `DataType` does not know the nullability but `Field` does. Thanks for reminding me! @lidavidm